Add IoT Hub Samples for Twin, Telemetry, C2D, and Methods (#648)
Azure/azure-sdk-for-c: sdk/iot/hub/samples/README.md

---
page_type: sample
languages:
  - c
products:
  - azure
  - azure-iot
  - azure-iot-hub
urlFragment: iot-hub-samples
---

# Azure IoT Hub Sample

This document explains samples for the Azure Embedded C SDK IoT Hub Client and how to use them.

## Key concepts

Key concepts are explained in detail [here][SDK_README_KEY_CONCEPTS].

## Samples for Azure IoT Hub Client APIs

This document describes how to use samples and what is done in each sample.

## Getting started

Getting started explained in detail [here][SDK_README_GETTING_STARTED].

## Examples

The following section documents various examples. All of them use the [Eclipse Paho MQTT C client][Eclipse_Paho].

> 1. [IoT Hub C2D][paho_c2d_sample]: Receive and view incoming C2D messages using the IoT Hub client.

Author: I went with alphabetical/the way the directory structure is set up. It's a toss up either way. Might let other clues in (I know it's a nit)?

> 1. [IoT Hub C2D][paho_c2d_sample]: Receive and view incoming C2D messages using the IoT Hub client.

Reviewer: Basic SKUs don't support the other features; Telemetry is the most important feature (the rest may or may not be required by devices). Reading the doc, users should find the most important features at top.
I would argue that C2D be moved last: most people should use Methods and Twin instead of C2D.

> 1. [IoT Hub C2D][paho_c2d_sample]: Receive and view incoming C2D messages using the IoT Hub client.

Author: reordered

---
page_type: sample
languages:
  - c
products:
  - azure
  - azure-iot
  - azure-iot-hub
urlFragment: iot-hub-samples

> ---

Reviewer 2: Remove this preamble? Is this meant to act as hidden metadata? It affects readability of the markdown on GitHub:
![image](https://user-images.githubusercontent.com/6527137/80770489-4fe92c00-8b05-11ea-91a9-725954fc928d.png)

Author: odd. I think it is this way across all of our stuff. I will likely leave it here for now but make a note to fix across the board if necessary.

# Azure IoT Hub Sample

> This document explains samples for the Azure Embedded C SDK IoT Hub Client and how to use them.

Reviewer 3: ... Azure SDK for Embedded C ...

Author: updated

## Key concepts

Key concepts are explained in detail [here][SDK_README_KEY_CONCEPTS].

## Samples for Azure IoT Hub Client APIs

This document describes how to use samples and what is done in each sample.

## Getting started

Getting started explained in detail [here][SDK_README_GETTING_STARTED].

## Examples

The following section documents various examples. All of them use the [Eclipse Paho MQTT C client][Eclipse_Paho].

### [IoT Hub C2D][c2d_sample]

> Receive and view incoming C2D messages using the IoT Hub client.

Reviewer 2: nit: consistency
```suggestion
Receive and view incoming C2D messages using the IoT Hub Client.
```

> ### [IoT Hub Methods][methods_sample]

Reviewer 2: This link is 404.

Invoke methods from the cloud. The sample supports a method named "double"
which will return back to you the value sent as the payload. The payload must be of the form:
```json
{
  "value": 10
}
```

> where 10 can be substituted for any number between `UINT64_T_MIN / 2` and `UINT64_T_MAX / 2`.

Reviewer 2: Isn't UINT64_T_MIN just 0? Was this meant to be INT64_T? Also, consider using the names from `limits.h`
https://www.tutorialspoint.com/c_standard_library/limits_h.htm

I don't get much search hit for `UINT64_T_MAX`.

```suggestion
where 10 can be substituted for any number between `0 and `ULONG_MAX / 2`.
```

Author: yes it should be `INT32` thank you

### [IoT Hub Telemetry][telemetry_sample]

> Send 5 telemetry messages using the IoT Hub client.

Reviewer 2: Same consistency nit here. Should the `c` in `client` be capital everywhere or lower case?
```suggestion
Send 5 telemetry messages using the IoT Hub Client.
```

### [IoT Hub Twin][twin_sample]
Use twin features such as updating reported properties, receiving the twin document, and receiving desired properties using the IoT Hub client.

## Troubleshooting

When interacting with the Azure IoT Hub using this C client library, errors are documented within the [MQTT State Machine][error_codes] requests.

## Next steps

> Start using the IoT Hub Client in your solutions. Our SDK details can be found at [SDK README][IOT_CLIENT_README].

Reviewer 2: ```suggestion
Start using the IoT Hub Client in your solutions. More details can be found in the [IOT SDK README][IOT_CLIENT_README].
```

### Additional Documentation

For extensive documentation on Azure IoT Hub, see the [API reference documentation][iot_hub_mqtt].

## Contributing

> This project welcomes contributions and suggestions. Find [more contributing][SDK_README_CONTRIBUTING] details here.

Reviewer 2: Is this section needed here?

Author: I'd say it couldn't hurt 🤷‍♂️

Reviewer: Please keep readme.md files as short as possible. I think we already have a contributing guide in root.

Author: removed

<!-- LINKS -->
[IOT_CLIENT_README]: ../../README.md
[SDK_README_CONTRIBUTING]:../../README.md#contributing
[SDK_README_GETTING_STARTED]: ../../README.md#getting-started
[SDK_README_KEY_CONCEPTS]: ../../README.md#key-concepts
[c2d_sample]: c2d/src/iot_hub_c2d_example.c

> [methods_sample]: methods/src/iot_methods_example.c

Reviewer 2: I think this needs to be updated to match file name: `iot_hub_methods_example.c`

Author: yup fixed thank you
